In [1]:
!pip install geocoder
!pip install folium

In [2]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print("All Required Libraries Imported!")

All Required Libraries Imported!


In [8]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
extracting_data = requests.get(url).text
wiki_data = BeautifulSoup(extracting_data, 'lxml')

In [4]:
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

content = wiki_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,'Borough': borough,'Neighborhood': neighborhood},ignore_index=True)

In [5]:
# clean dataframe 
toronto = toronto[toronto.Borough!='Not assigned']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1

In [9]:
df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head()

Postalcode         Borough                            Neighborhood
0      M1A\n  Not assigned\n                          Not assigned\n
1      M1B\n   Scarborough\n                          Malvern, Rouge
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill
4      M1G\n   Scarborough\n                                  Woburn

In [10]:
df.describe()

Postalcode         Borough    Neighborhood
count         180             180             180
unique        180              11             100
top         M7K\n  Not assigned\n  Not assigned\n
freq            1              77              77

In [12]:
df = df.dropna()
empty = 'Not assigned'
df = df[(df.Postalcode != empty ) & (df.Borough != empty) & (df.Neighborhood != empty)]

In [13]:
df.head()

Postalcode         Borough                            Neighborhood
0      M1A\n  Not assigned\n                          Not assigned\n
1      M1B\n   Scarborough\n                          Malvern, Rouge
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill
4      M1G\n   Scarborough\n                                  Woburn

In [14]:
def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df_2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')

In [15]:
df_2.describe()

Postalcode         Borough    Neighborhood
count         180             180             180
unique        180              11             100
top         M7K\n  Not assigned\n  Not assigned\n
freq            1              77              77

In [16]:
print(df_2.shape)
df_2.head()

(180, 3)


Postalcode         Borough                            Neighborhood
0      M1A\n  Not assigned\n                          Not assigned\n
1      M1B\n   Scarborough\n                          Malvern, Rouge
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill
4      M1G\n   Scarborough\n                                  Woburn

In [17]:
def get_latilong(postal_code):
    lati_long_coords = None
    while(lati_long_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(postal_code))
        lati_long_coords = g.latlng
    return lati_long_coords
    
get_latilong('M4G')

[43.70941386000004, -79.36309957799995]

In [18]:
# Retrieving Postal Code Co-ordinates
postal_codes = df_2['Postalcode']    
coords = [ get_latilong(postal_code) for postal_code in postal_codes.tolist() ]

In [19]:
# Adding Columns Latitude & Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])
df_2['Latitude'] = df_coords['Latitude']
df_2['Longitude'] = df_coords['Longitude']

In [25]:
df_2[df_2.Postalcode == 'M4G']

Empty DataFrame
Columns: [Postalcode, Borough, Neighborhood, Latitude, Longitude]
Index: []

In [22]:
df_2.head(10)

Postalcode         Borough                                     Neighborhood  \
0      M1A\n  Not assigned\n                                   Not assigned\n   
1      M1B\n   Scarborough\n                                   Malvern, Rouge   
2      M1C\n   Scarborough\n           Rouge Hill, Port Union, Highland Creek   
3      M1E\n   Scarborough\n                Guildwood, Morningside, West Hill   
4      M1G\n   Scarborough\n                                           Woburn   
5      M1H\n   Scarborough\n                                        Cedarbrae   
6      M1J\n   Scarborough\n                              Scarborough Village   
7      M1K\n   Scarborough\n      Kennedy Park, Ionview, East Birchmount Park   
8      M1L\n   Scarborough\n                  Golden Mile, Clairlea, Oakridge   
9      M1M\n   Scarborough\n  Cliffside, Cliffcrest, Scarborough Village West   

    Latitude  Longitude  
0  43.648690 -79.385440  
1  43.808626 -79.189913  
2  43.785779 -79.157368  
3  43.765806 -79.185284  
4  43.771545 -79.218135  
5  43.768791 -79.238813  
6  43.744203 -79.228725  
7  43.726881 -79.265694  
8  43.713340 -79.284942  
9  43.723538 -79.228353

In [26]:
address = 'Scarborough,Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude_x = location.latitude
longitude_y = location.longitude
print('The Geograpical Co-ordinate of Seattle,Washington are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Seattle,Washington are 43.773077, -79.257774.


In [27]:
map_Scarborough = folium.Map(location=[latitude_x, longitude_y], zoom_start=10)

for lat, lng, nei in zip(df_2['Latitude'], df_2['Longitude'], df_2['Neighborhood']):
    
    label = '{}'.format(nei)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Scarborough)  
    
map_Scarborough

In [28]:
address = 'Scarborough,Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude_n1 = location.latitude
longitude_n1 = location.longitude
print('The Geograpical Co-ordinate of Neighborhood_1 are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Neighborhood_1 are 43.773077, -79.257774.


In [29]:
 #@hiddel_cell
CLIENT_ID = 'DPBYY4JUY3DU20ALPSUV4ONY2K1GOJJKJ1NIHBB32XEMOVYY' # my Foursquare ID
CLIENT_SECRET = '1MV443TYEP4HUO0WDUW5NQ5W10L2Y4G05NWG11WIR3NUGC5B' # my Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: '+CLIENT_ID)
print('CLIENT_SECRET: '+CLIENT_SECRET)

Your credentails:
CLIENT_ID: DPBYY4JUY3DU20ALPSUV4ONY2K1GOJJKJ1NIHBB32XEMOVYY
CLIENT_SECRET: 1MV443TYEP4HUO0WDUW5NQ5W10L2Y4G05NWG11WIR3NUGC5B


In [30]:
radius = 700 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude_n1, 
   longitude_n1, 
    radius, 
   LIMIT)
results = requests.get(url).json()

In [31]:
venues=results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)
nearby_venues.columns

Index(['reasons.count', 'reasons.items', 'referralId', 'venue.categories',
       'venue.id', 'venue.location.address', 'venue.location.cc',
       'venue.location.city', 'venue.location.country',
       'venue.location.crossStreet', 'venue.location.distance',
       'venue.location.formattedAddress', 'venue.location.labeledLatLngs',
       'venue.location.lat', 'venue.location.lng',
       'venue.location.neighborhood', 'venue.location.postalCode',
       'venue.location.state', 'venue.name', 'venue.photos.count',
       'venue.photos.groups', 'venue.venuePage.id'],
      dtype='object')

In [32]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [33]:
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()

venue.name  \
0               Disney Store   
1   St. Andrews Fish & Chips   
2                    SEPHORA   
3                  DAVIDsTEA   
4  American Eagle Outfitters   

                                    venue.categories  venue.location.lat  \
0  [{'id': '4bf58dd8d48988d1f3941735', 'name': 'T...           43.775537   
1  [{'id': '4edd64a0c7ddd24ca188df1a', 'name': 'F...           43.771865   
2  [{'id': '4bf58dd8d48988d10c951735', 'name': 'C...           43.775017   
3  [{'id': '4bf58dd8d48988d1dc931735', 'name': 'T...           43.776320   
4  [{'id': '4bf58dd8d48988d103951735', 'name': 'C...           43.776012   

   venue.location.lng  
0          -79.256833  
1          -79.252645  
2          -79.258109  
3          -79.258688  
4          -79.258334

In [34]:
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(5)

name         categories        lat        lng
0               Disney Store   Toy / Game Store  43.775537 -79.256833
1   St. Andrews Fish & Chips  Fish & Chips Shop  43.771865 -79.252645
2                    SEPHORA     Cosmetics Shop  43.775017 -79.258109
3                  DAVIDsTEA           Tea Room  43.776320 -79.258688
4  American Eagle Outfitters     Clothing Store  43.776012 -79.258334

In [35]:
# Top 10 Categories
a=pd.Series(nearby_venues.categories)
a.value_counts()[:10]

Clothing Store        7
Coffee Shop           4
Restaurant            4
Tea Room              2
Sandwich Place        2
Pharmacy              2
Gas Station           2
Cosmetics Shop        1
Bar                   1
Italian Restaurant    1
Name: categories, dtype: int64

In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=700):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # making GET request
        venue_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [37]:
# Nearby Venues
Scarborough_venues = getNearbyVenues(names=df_2['Neighborhood'],
                                   latitudes=df_2['Latitude'],
                                   longitudes=df_2['Longitude']
                                  )

Not assigned

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Parkwoods
Don Mills
Don Mills
Not assigned

Not assigned

Bathurst Manor, Wilson Heights, Downsview North
No

In [38]:
print('There are {} Uniques Categories.'.format(len(Scarborough_venues['Venue Category'].unique())))
Scarborough_venues.groupby('Neighborhood').count().head()

There are 309 Uniques Categories.


Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                           31   
Alderwood, Long Branch                                              10   
Bathurst Manor, Wilson Heights, Downsview North                     25   
Bayview Village                                                      5   
Bedford Park, Lawrence Manor East                                   24   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                            31   
Alderwood, Long Branch                                               10   
Bathurst Manor, Wilson Heights, Downsview North                      25   
Bayview Village                                                       5   
Bedford Park, Lawrence Manor East                                    24   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                           31              31   
Alderwood, Long Branch                              10              10   
Bathurst Manor, Wilson Heights, Downsview North     25              25   
Bayview Village                                      5               5   
Bedford Park, Lawrence Manor East                   24              24   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                     31   
Alderwood, Long Branch                                        10   
Bathurst Manor, Wilson Heights, Downsview North               25   
Bayview Village                                                5   
Bedford Park, Lawrence Manor East                             24   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                    31  
Alderwood, Long Branch                                       10  
Bathurst Manor, Wilson Heights, Downsview North              25  
Bayview Village                                               5  
Bedford Park, Lawrence Manor East                            24

In [39]:
# one hot encoding
Scarborough_onehot = pd.get_dummies(Scarborough_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Scarborough_onehot['Neighborhood'] = Scarborough_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Scarborough_onehot.columns[-1]] + list(Scarborough_onehot.columns[:-1])
Scarborough_onehot = Scarborough_onehot[fixed_columns]
Scarborough_grouped = Scarborough_onehot.groupby('Neighborhood').mean().reset_index()
Scarborough_onehot.head(5)

Yoga Studio  Accessories Store  Afghan Restaurant  African Restaurant  \
0            0                  0                  0                   0   
1            0                  0                  0                   0   
2            0                  0                  0                   0   
3            0                  0                  0                   0   
4            0                  0                  0                   0   

   Airport  American Restaurant  Antique Shop  Aquarium  Arcade  Art Gallery  \
0        0                    0             0         0       0            0   
1        0                    0             0         0       0            0   
2        0                    0             0         0       0            0   
3        0                    0             0         0       0            0   
4        0                    0             0         0       0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Dealership  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                0          0           0           0       0     0    0   
1                0          0           0           0       0     0    0   
2                0          0           0           0       0     0    0   
3                0          0           0           0       0     0    0   
4                0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  Big Box Store  \
0      0          0         0           0                   0              0   
1      0          0         0           0                   0              0   
2      0          0         0           0                   0              0   
3      0          0         0           0                   0              0   
4      0          0         0           0                   0              0   

   Bike Shop  Bistro  Bookstore  Botanical Garden  Boutique  \
0          0       0          0                 0         0   
1          0       0          0                 0         0   
2          0       0          0                 0         0   
3          0       0          0                 0         0   
4          0       0          0                 0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  Bridge  \
0                     0               0        0            0       0   
1                     0               0        0            0       0   
2                     0               0        0            0       0   
3                     0               0        0            0       0   
4                     0               0        0            0       0   

   Bubble Tea Shop  Buffet  Building  Burger Joint  Burrito Place  Bus Line  \
0                0       0         0             0              0         0   
1                0       0         0             0              0         0   
2                0       0         0             0              0         0   
3                0       0         0             0              0        

In [40]:
num_top_venues = 5
for hood in Scarborough_grouped['Neighborhood']:
    print("---- "+hood+" ----")
    temp =Scarborough_grouped[Scarborough_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Agincourt ----
                venue  freq
0  Chinese Restaurant  0.13
1       Shopping Mall  0.06
2    Department Store  0.03
3  Seafood Restaurant  0.03
4                Bank  0.03


---- Alderwood, Long Branch ----
               venue  freq
0  Convenience Store   0.1
1        Pizza Place   0.1
2     Sandwich Place   0.1
3                Gym   0.1
4        Coffee Shop   0.1


---- Bathurst Manor, Wilson Heights, Downsview North ----
                 venue  freq
0                 Park  0.08
1          Coffee Shop  0.08
2                 Bank  0.08
3       Ice Cream Shop  0.04
4  Fried Chicken Joint  0.04


---- Bayview Village ----
                        venue  freq
0                        Park   0.6
1  Construction & Landscaping   0.2
2                       Trail   0.2
3                Optical Shop   0.0
4                      Office   0.0


---- Bedford Park, Lawrence Manor East ----
                venue  freq
0          Restaurant  0.08
1      Sandwich Place  0.08
2      

In [41]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [42]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Scarborough_grouped['Neighborhood']

for ind in np.arange(Scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch           Coffee Shop   
2  Bathurst Manor, Wilson Heights, Downsview North                  Park   
3                                  Bayview Village                  Park   
4                Bedford Park, Lawrence Manor East            Restaurant   

        2nd Most Common Venue 3rd Most Common Venue  \
0               Shopping Mall       Bubble Tea Shop   
1              Sandwich Place     Convenience Store   
2                 Coffee Shop                  Bank   
3  Construction & Landscaping                 Trail   
4              Sandwich Place    Italian Restaurant   

       4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0  Latin American Restaurant                Bakery                  Bank   
1                Pizza Place           Gas Station                   Pub   
2                Pizza Place            Restaurant         Shopping Mall   
3              Women's Store    Falafel Restaurant      Doner Restaurant   
4                Coffee Shop     Indian Restaurant               Butcher   

  7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
0   Japanese Restaurant      Sushi Restaurant                  Supermarket   
1              Pharmacy          Dance Studio                          Gym   
2           Bridal Shop      Community Center    Middle Eastern Restaurant   
3            Donut Shop   Dumpling Restaurant  Eastern European Restaurant   
4                  Café          Liquor Store                  Sports Club   

  10th Most Common Venue  
0         Breakfast Spot  
1     Athletics & Sports  
2      Mobile Phone Shop  
3      Electronics Store  
4       Sushi Restaurant

In [43]:
# Using K-Means to cluster neighborhood into 3 clusters
Scarborough_grouped_clustering = Scarborough_grouped.drop('Neighborhood', 1)
kmeans = KMeans(n_clusters=3, random_state=0).fit(Scarborough_grouped_clustering)
kmeans.labels_

array([0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 2, 0, 1, 0], dtype=int32)

In [44]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Scarborough_merged =df_2.iloc[:16,:]

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Scarborough_merged = Scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Scarborough_merged.head()# check the last columns!

Postalcode         Borough                            Neighborhood  \
0      M1A\n  Not assigned\n                          Not assigned\n   
1      M1B\n   Scarborough\n                          Malvern, Rouge   
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek   
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill   
4      M1G\n   Scarborough\n                                  Woburn   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.648690 -79.385440               0           Coffee Shop   
1  43.808626 -79.189913               1                  Park   
2  43.785779 -79.157368               1                   Bar   
3  43.765806 -79.185284               0           Pizza Place   
4  43.771545 -79.218135               2           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                 Hotel            Restaurant   
1                 Trail         Women's Store              Dive Bar   
2     Fish & Chips Shop                  Park    Falafel Restaurant   
3      Greek Restaurant  Fast Food Restaurant                  Park   
4                  Park      Business Service         Women's Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Japanese Restaurant          Concert Hall         Movie Theater   
1      Doner Restaurant            Donut Shop   Dumpling Restaurant   
2      Doner Restaurant            Donut Shop   Dumpling Restaurant   
3           Coffee Shop                  Bank            Restaurant   
4               Dog Run            Donut Shop   Dumpling Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          American Restaurant              Beer Bar              Bookstore  
1  Eastern European Restaurant     Electronics Store   Ethiopian Restaurant  
2  Eastern European Restaurant     Electronics Store   Ethiopian Restaurant  
3               Medical Center          Intersection    Rental Car Location  
4  Eastern European Restaurant     Electronics Store   Ethiopian Restaurant

In [45]:
kclusters = 10

In [46]:
# create map
map_clusters = folium.Map(location=[latitude_x, longitude_y], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]
print(rainbow)
# add markers to the map

markers_colors = []
for lat, lon, nei , cluster in zip(Scarborough_merged['Latitude'], 
                                   Scarborough_merged['Longitude'], 
                                   Scarborough_merged['Neighborhood'], 
                                   Scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(nei) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

['#8000ff', '#4856fb', '#10a2f0', '#2adddd', '#62fbc4', '#9cfba4', '#d4dd80', '#ffa256', '#ff562c', '#ff0000']


In [47]:
df1=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 0,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df2=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 1,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df3=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 2,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]

In [48]:
Scarborough_Avg_HousingPrice=pd.DataFrame({"Neighborhood":df_2["Neighborhood"],
                                      "Average_Housing_Price":[335000.0,286600.0,175000.0,225900.0,219400.0,
                                                               573900.0,225000.0,370500.0,370500.0,433500.0,279200.0,
                                                               279200.0,225000.0,370500.,255400.0,433500.0,433500.0,
                                                               435000.0,289500.0,265000.0,285900.0,239400.0,
                                                               589900.0,295000.0,380500.0,378500.0,438500.0,229200.0,
                                                               229200.0,365000.0,388500.,285400.0,493500.0,477500.0,378000.0,316600.0,195000.0,225900.0,219400.0,
                                                               573900.0,367000.0,370500.0,370500.0,363500.0,279200.0,
                                                               279200.0,271000.0,370500.,255400.0,383500.0,433500.0,335000.0,286600.0,185000.0,225900.0,219400.0,
                                                               573900.0,329000.0,370500.0,370500.0,533500.0,279200.0,
                                                               279200.0,375000.0,370500.,255400.0,493500.0,433500.0,335000.0,286600.0,165000.0,225900.0,219400.0,
                                                               573900.0,425000.0,370500.0,370500.0,433500.0,279200.0,
                                                               279200.0,195000.0,370500.,255400.0,403500.0,433500.0,335000.0,286600.0,187000.0,225900.0,219400.0,
                                                               573900.0,325000.0,370500.0,370500.0,333500.0,279200.0,
                                                               279200.0,289000.0,370500.,255400.0,413500.0,433500.0,254800.0
                                                              
                                                              
                                                              ]})

ValueError: array length 103 does not match index length 180

In [49]:
Scarborough_Avg_HousingPrice.set_index('Neighborhood',inplace=True,drop=True)

NameError: name 'Scarborough_Avg_HousingPrice' is not defined

In [50]:
Scarborough_Avg_HousingPrice.plot(kind='bar',figsize=(24,18),alpha=0.75)

NameError: name 'Scarborough_Avg_HousingPrice' is not defined

In [51]:
clusters=pd.DataFrame({"Cluster1":df1["Neighborhood"],
                      "Cluster2":df2["Neighborhood"],
                      "Cluster4":df3["Neighborhood"]})
clusters = clusters.replace(np.nan, '', regex=True)

In [52]:
new_Scarborough=Scarborough_merged.set_index("Neighborhood",drop=True)
#Source:https://www.greatschools.org
Scarborough_school_ratings=pd.DataFrame({"Neighborhood":df["Neighborhood"],
                                      "Top School Rating":[7,9,5,8,10,10,7,10,1,2,1,2,7,2,3,2,6,
                                                           5,4,8,9,9,6,6,4,5,4,6,8,10,8,9,6,2,
                                                           10,2,5,8,9,6,6,10,8,9,1,2,3,4,5,6,9,
                                                           8,5,9,6,9,6,4,8,10,2,5,6,3,9,8,7,
                                                           7,8,5,8,9,1,5,4,7,2,3,6,6,9,4,8,7,
                                                           4,8,9,2,6,4,7,5,10,4,6,8,9,7,5,6,5,8,7
                                                           ]})

ValueError: array length 103 does not match index length 180

In [53]:
Scarborough_school_ratings.set_index('Neighborhood',inplace=True,drop=True)

NameError: name 'Scarborough_school_ratings' is not defined

In [54]:
Scarborough_school_ratings.plot(kind='bar',figsize=(16,10),color='green',alpha=0.75);

NameError: name 'Scarborough_school_ratings' is not defined

Conclusion: In this project, using k-means cluster algorithm I separated the neighborhood into 10(Ten) different clusters and for 103 different lattitude and logitude from dataset, which have very-similar neighborhoods around them. Using the charts above results presented to a particular neighborhood based on average house prices and school rating have been made.

Blog Post Link:

My email address